### PR TITLE
FIX: Sort conditions in bids_gen_info to ensure consistent order

### PR DESCRIPTION
**Pull request by TheChymera**
@effigies 
If this is not done, condition order for orthogonalizarion is
unpredictable.

## Acknowledgment

- [x] \(Mandatory\) I acknowledge that this contribution will be available under the Apache 2 license.

### CODE REVIEW

File: nipype/algorithms/modelgen.py [hunk 1]
```diff
@@ -176,6 +176,7 @@ def bids_gen_info(bids_event_files,
             f_events = csv.DictReader(f, skipinitialspace=True, delimiter='\t')
             events = [{k: v for k, v in row.items()} for row in f_events]
         conditions = list(set([i[condition_column] for i in events]))
```
Comment (effigies) [inline nipype/algorithms/modelgen.py]: Just to condense this into one line:

```suggestion
        conditions = sorted(set([i[condition_column] for i in events]))
```

**Comment by satra** [comment]
@effigies  and @TheChymera - yes, it may be useful to see a point of failure where this will not work. i'm mostly concerned whether this worked and stopped working or did not work properly ever.

also whether this works across both FSL and SPM.

**Comment by effigies** [comment]
This was recently added (#2845) as a BIDS-y alternative to the existing events files, so it's not a failure of a historical construct.

**Comment by effigies** [comment]
@TheChymera Can you review the suggestion? And is there any chance of a small regression test?

**Comment by effigies** [comment]
@TheChymera Just a note that the 1.1.9 release is coming up on Monday, and it would be good to have any PRs we want to include in on Friday.

**Comment by effigies** [comment]
In this case, a simple regression test would be a BIDS-style event file and a function that compares the output of `bids_gen_info` to an expected output:

```Python
def test_bids_gen_info():
    fname = ...
    res = bids_gen_info(fname)
    # Tests of properties
    assert len(res) == 1
    assert res[0].onsets == ...
    assert res[0].durations == ...
    assert res[0].amplitudes == ...
    assert res[0].conditions == ...
```

**Comment by TheChymera** [comment]
@effigies what directory should I put the example file into? I can't find other example files, e.g. `test1.nii` anywhere.

**Comment by satra** [comment]
@TheChymera - if you create a test with:

```python
def test_bids_info(tmpdir):
    filename = tmpdir / 'testfile.nii'
    import nibabel as nb
    # create a local file here and use it for a test.
    ....
    img.to_nifti(filename)
    # use filename for test
```

**Comment by TheChymera** [comment]
@satra so I should create the .tsv file as part of the test? that's a lot more and a lot less legible text than if I tracked it somewhere. But where? `examples/`? In the test dir?

**Comment by satra** [comment]
the general place where test data go is `testing/data` but most files there are empty. so you can put small files, but not large ones.

**Comment by TheChymera** [comment]
@satra @effigies works over here, hope this is good now :)

**Comment by TheChymera** [comment]
@satra @effigies it works on my box locally, but I can't get the path to work on circleci. Any advice?

File: nipype/algorithms/tests/test_modelgen.py [hunk 2]
```diff
@@ -12,10 +12,18 @@
...
 import numpy.testing as npt
 from nipype.interfaces.base import Bunch, TraitError
-from nipype.algorithms.modelgen import (SpecifyModel, SpecifySparseModel,
-                                        SpecifySPMModel)
+from nipype.algorithms.modelgen import (bids_gen_info, SpecifyModel,
+                                        SpecifySparseModel, SpecifySPMModel)
 
 
+def test_bids_gen_info():
+    fname = os.path.realpath('../../testing/data/events.tsv')
```
Comment (satra) [inline nipype/algorithms/tests/test_modelgen.py]:
```suggestion
    fname = os.path.realpath('data/events.tsv')
```

Comment (TheChymera) [reply nipype/algorithms/tests/test_modelgen.py]: ah, I see I had been running the tests from another directory..

Comment (TheChymera) [reply nipype/algorithms/tests/test_modelgen.py]: I still get
```
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/pytest-of-neuro/pytest-0/popen-gw0/test_warning0/data/events.tsv'
```

File: nipype/algorithms/tests/test_modelgen.py [hunk 3]
```diff
@@ -12,10 +12,18 @@
...
 import numpy.testing as npt
 from nipype.interfaces.base import Bunch, TraitError
-from nipype.algorithms.modelgen import (SpecifyModel, SpecifySparseModel,
-                                        SpecifySPMModel)
+from nipype.algorithms.modelgen import (bids_gen_info, SpecifyModel,
+                                        SpecifySparseModel, SpecifySPMModel)
 
 
+def test_bids_gen_info():
+    fname = os.path.realpath('data/events.tsv')
```
Comment (satra) [inline nipype/algorithms/tests/test_modelgen.py]:
```suggestion
    fname = npt.example_data('events.tsv')
```

Comment (satra) [reply nipype/algorithms/tests/test_modelgen.py]: sorry i forgot that location is not a default in the regular tests.

File: nipype/algorithms/modelgen.py [hunk 4]
```diff
@@ -175,7 +175,11 @@ def bids_gen_info(bids_event_files,
         with open(bids_event_file) as f:
             f_events = csv.DictReader(f, skipinitialspace=True, delimiter='\t')
             events = [{k: v for k, v in row.items()} for row in f_events]
-        conditions = list(set([i[condition_column] for i in events]))
+        if not condition_column:
+            condition_column = '_trial_type'
+            for i in events:
+                i.update({condition_column:'ev0'})
```
Comment (effigies) [inline nipype/algorithms/modelgen.py]:
```suggestion
                i.update({condition_column: 'ev0'})
```

File: nipype/algorithms/modelgen.py [hunk 5]
```diff
@@ -199,7 +200,6 @@ def bids_gen_info(bids_event_files,
         info.append(runinfo)
     return info
 
-
```
Comment (effigies) [inline nipype/algorithms/modelgen.py]: Style guides want two newlines between functions.

File: nipype/algorithms/tests/test_modelgen.py (hunk 3 above)
Comment (TheChymera) [reply nipype/algorithms/tests/test_modelgen.py]: @satra still not working :-/

**Comment by TheChymera** [comment]
Added separate import because @satra's suggestion failed with

```
def test_bids_gen_info():
>       fname = npt.example_data('events.tsv')
E       AttributeError: module 'numpy.testing' has no attribute 'example_data'

/src/nipype/nipype/algorithms/tests/test_modelgen.py:20: AttributeError
```

**Comment by TheChymera** [comment]
Ok, can't reproduce this new failure locally, sounds like some unicode support is missing from the testing environment:

```
def test_bids_gen_info():
        fname = example_data('events.tsv')
>       res = bids_gen_info([fname])

/src/nipype/nipype/algorithms/tests/test_modelgen.py:22: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/src/nipype/nipype/algorithms/modelgen.py:177: in bids_gen_info
    f_events = csv.DictReader(f, skipinitialspace=True, delimiter='\t')
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <csv.DictReader instance at 0x7f4a4d694d88>
f = <closed file '/src/nipype/nipype/testing/data/events.tsv', mode 'r' at 0x7f4a4d943300>
fieldnames = None, restkey = None, restval = None, dialect = 'excel', args = ()
kwds = {'delimiter': '	', 'skipinitialspace': True}

    def __init__(self, f, fieldnames=None, restkey=None, restval=None,
                 dialect="excel", *args, **kwds):
        self._fieldnames = fieldnames   # list of keys for the dict
        self.restkey = restkey          # key to catch long rows
        self.restval = restval          # default value for short rows
>       self.reader = reader(f, dialect, *args, **kwds)
E       TypeError: "delimiter" must be string, not unicode

/opt/miniconda-latest/envs/neuro/lib/python2.7/csv.py:79: TypeError
```

File: nipype/algorithms/modelgen.py [hunk 6]
```diff
@@ -175,7 +176,11 @@ def bids_gen_info(bids_event_files,
         with open(bids_event_file) as f:
             f_events = csv.DictReader(f, skipinitialspace=True, delimiter='\t')
```
Comment (effigies) [inline nipype/algorithms/modelgen.py]: This should resolve the Python 2/Unicode issue.

```suggestion
            f_events = csv.DictReader(f, skipinitialspace=True, delimiter=str('\t'))
```

**Comment by TheChymera** [comment]
@effigies 

ok, so using the str type from builtins gives me this:

```
E       TypeError: "delimiter" must be string, not newstr
```

The only way I was able to fix it is by re-assigning the base type before the import, since removing the import breaks another function.

**Comment by effigies** [comment]
Going to go ahead and merge. Thanks for your patience!

**Pull request merged**